Add basic GATs reference information (#1265)
rust-lang/reference: src/items/associated-items.md

default implementation in traits.

An *associated type declaration* declares a signature for associated type
definitions. It is written in one of the following forms, where `Assoc` is the
name of the associated type, `Params` is a comma-separated list of type,
lifetime or const parameters, `Bounds` is a plus-separated list of trait bounds
on the associated type, and `WhereBounds` is a comma-separated list of bounds on
parameters:

> ```rust,ignore

Reviewer: Can you include a comment explaining why it is ignored? We try to avoid `ignore` as much as possible.

```suggestion
<!-- ignore: illustrative example forms -->
```rust,ignore
```

type Assoc;
type Assoc: Bounds;
type Assoc<Params>;

> type Assoc<Params>: Bounds;

Reviewer 2: It occurs to me that it would be worth clarifying the distinction between `Bounds` and `WhereBounds`!

Author: That is stated in the paragraph above? Should we be more detailed?

Reviewer 2: I think we could be more explicit. Left some suggestions. WDYT?

type Assoc<Params> where WhereBounds;
type Assoc<Params>: Bounds where WhereBounds;
```

The identifier is the name of the declared type alias. The optional trait bounds
must be fulfilled by the implementations of the type alias.
There is an implicit [`Sized`] bound on associated types that can be relaxed using the special `?Sized` bound.

> An *associated type definition* defines a type alias on for the implementation

Reviewer 3: Small copy note: "on for" isn't grammatical here, probably just meant to pick one.

Reviewer 2: good catch

> An *associated type definition* defines a type alias on for the implementation

Reviewer 2: ```suggestion
An *associated type definition* defines a type alias for the implementation
```

of a trait on a type. They are written similarly to an *associated type declaration*,
but cannot contain `Bounds`, but instead must contain a `Type`:

> ```rust,ignore

Reviewer: ```suggestion
<!-- ignore: illustrative example forms -->
```rust,ignore
```

type Assoc = Type;

> type Assoc<Params> = Type<Params>;

Reviewer 2: ```suggestion
type Assoc<Params> = Type;
```

This is inconsistent with the other examples, no?

Author: Yeah, I specifically wanted to highlight that you can use those `Params` in the defined type. But maybe it's just too confusing

Reviewer 2: Yeah, I can see that. I could also see people interpreting it to mean that the type *has* to have that form (e.g., `Vec<T>` or something).

Reviewer 2: Could do 

```rust
type Assoc<Params> = Type; // the type `Type` here may reference `Params`
```

if you want to make it extra clear.

> type Assoc<Params> where WhereBounds = Type;
> type Assoc<Params> = Type where WhereBounds;

Reviewer 2: ```suggestion
type Assoc<Params> = Type where WhereBounds;
type Assoc<Params> where WhereBounds = Type; // deprecated, prefer the form above
```

Nit, but it makes sense to let people know that this where clause placement is not the one to use.

default implementation in traits.

An *associated type declaration* declares a signature for associated type
definitions. It is written in one of the following forms, where `Assoc` is the
name of the associated type, `Params` is a comma-separated list of type,

> lifetime or const parameters, `Bounds` is a plus-separated list of trait bounds
> on the associated type, and `WhereBounds` is a comma-separated list of bounds on
> parameters:

Reviewer 2: ```suggestion
lifetime or const parameters, `Bounds` is a plus-separated list of trait bounds
that the associated type must meet, and `WhereBounds` is a comma-separated list of bounds that the parameters must meet:
```

<!-- ignore: illustrative example forms -->
```rust,ignore
type Assoc;
type Assoc: Bounds;
type Assoc<Params>;
type Assoc<Params>: Bounds;
type Assoc<Params> where WhereBounds;
type Assoc<Params>: Bounds where WhereBounds;

> ```

Reviewer 2: Maybe something like

### Relationship between `Bounds` and `WhereBounds`

In this example:

```rust
trait Example {
    type Output<T>: Ord where T: Debug
}
```

Given a reference to the associated type like `<X as Example>::Output<Y>`, the associated type itself must be `Ord`, and the type `Y` must be `Debug`.

associated type definition. Furthermore, if `Item` is a type parameter, then
`Item::Assoc` can be used in type parameters.

> Associated types may include [generic parameters] or [where clauses]; these may

Reviewer 4: 'and', not 'or'

> be referred to generic associated types, or GATs. If the type `Thing` has an

Author: ```suggestion
be referred to as generic associated types, or GATs. If the type `Thing` has an
```

> be referred to generic associated types, or GATs. If the type `Thing` has an

Author: Thanks! Good catch.

> be referred to as generic associated types, or GATs. If the type `Thing` has an

Reviewer 5: ```suggestion
be referred to as *generic associated types*, or *GATs*. If the type `Thing` has an
```

I also think "may be referred to" should just say "these are called".

}
```

### Required where clauses on generic associated types

Generic associated type declarations on traits currently may require a list of

> where clauses, dependent on functions in the trait and how the GAT is used.

Reviewer 2: I feel we should indicate somewhere that these rules are "in flux". Perhaps we could link to [the page on the generic-associated-types initiative repository](https://rust-lang.github.io/generic-associated-types-initiative/explainer/required_bounds.html)? Alternatively, isn't there an issue that we redirect people to?

Author: good idea
